magento/magento2
# [GraphQL] Paypal PayFlowPro store the payment details in myaccount for future

### Description (*)

This PR provide ability to store my Paypal (PayFlowPro) payment method details in myaccount.

### Related Pull Requests
https://github.com/magento/partners-magento2-infrastructure/pull/15


**Useful links:**
[PayFloPro Documentation](https://docs.magento.com/user-guide/payment/paypal-payflow-pro.html)
[PayFloPro GraphQl Documentation](https://devdocs.magento.com/guides/v2.4/graphql/payment-methods/payflow-pro.html)
[How to create test PayPal accout for testing](https://developer.paypal.com/docs/payflow/test-hosted-pages/#create-a-business-account-in-the-sandbox)
[List of credit carts](https://developer.paypal.com/docs/payflow/payflow-pro/payflow-pro-testing/#credit-card-numbers-for-testing)

### Fixed Issues (if relevant)

1. Fixes magento/magento2#28520
2. https://github.com/magento/magento2/issues/28558

### Manual testing scenarios (*)

**1.** You should configure Payflowpro payment method.
**2.** Login as Customer 
**3.** Add product to cart and go to payment step on checkout
<img width="1439" alt="Screenshot 2020-06-22 at 13 05 15" src="https://user-images.githubusercontent.com/13517181/85275500-18558b00-b489-11ea-9f74-d316dd124eca.png">
**4.** Get customer quote id by running request
<pre>
{
  customerCart {
    id
    items {
      id
      product {
        name
        sku
      }
      quantity
    }
  }
}
</pre>
<img width="1378" alt="Screenshot 2020-06-22 at 13 10 35" src="https://user-images.githubusercontent.com/13517181/85275994-cd884300-b489-11ea-9837-4d5f9a1a9018.png">

**5.**  Set payment method for quote
<pre>
  setPaymentMethodOnCart(
    input: {
    cart_id: "zqYVjMInihkXlKVJAfgbY2yQsVdc6qLR"
    payment_method: {
      code: "payflowpro"
      payflowpro: {
           is_active_payment_token_enabler: true
          cc_details: {
           cc_type: "VI"
           cc_exp_year: 2025
           cc_exp_month: 7
           cc_last_4: 1881
        }
      }
    }
  })
  {
    cart {
      id
      selected_payment_method {
        code
        purchase_order_number
      }
  }
  }
}
</pre>
<img width="1304" alt="Screenshot 2020-06-22 at 13 14 32" src="https://user-images.githubusercontent.com/13517181/85276370-599a6a80-b48a-11ea-9b03-5c49e4745932.png">

**6.** Get secure tokens

<pre>
mutation {
  createPayflowProToken(
    input: {
      cart_id: "zqYVjMInihkXlKVJAfgbY2yQsVdc6qLR"
      urls: {
        return_url: "paypal/transparent/return"
        cancel_url: "paypal/transparent/cancel"
        error_url: "paypal/transparent/error"
      }
      
    }
  ) {
    response_message
    result
    result_code
    secure_token
    secure_token_id
  }
}
</pre>
<img width="1278" alt="Screenshot 2020-06-22 at 13 16 54" src="https://user-images.githubusercontent.com/13517181/85276640-d4638580-b48a-11ea-9f71-1e18bfdde0bf.png">

**7.** Run `handlePayflowProResponse` mutation 
[handlePayflowProResponse Documentation](https://devdocs.magento.com/guides/v2.4/graphql/mutations/handle-payflow-pro-response.html)
You can get `paypal_payload` parameter from 
<img width="1439" alt="Screenshot 2020-06-22 at 13 37 35" src="https://user-images.githubusercontent.com/13517181/85278543-c5ca9d80-b48d-11ea-9855-635cf4b53f24.png">


<pre>
mutation {
  handlePayflowProResponse(
    input: {
      cart_id: "zqYVjMInihkXlKVJAfgbY2yQsVdc6qLR"
      paypal_payload: "BILLTOCITY=dcdcd&AMT=0.00&BILLTOEMAIL=o.usik%40atwix.com&BILLTOSTREET=dcdc%40gmail.com&CC_TYPE=VI&SHIPTOCITY=dcdcd&NAMETOSHIP=Oleh+Usik&ZIP=12345&BILLTOLASTNAME=Usik&BILLTOFIRSTNAME=Oleh&RESPMSG=Verified&PROCCVV2=M&STATETOSHIP=AK&NAME=Oleh+Usik&BILLTOZIP=12345&CVV2MATCH=Y&PNREF=A71AADB57BB0&ZIPTOSHIP=12345&SHIPTOCOUNTRY=US&SHIPTOSTREET=dcdc%40gmail.com&CITY=dcdcd&HOSTCODE=A&LASTNAME=Usik&STATE=AK&SECURETOKEN=7dTXlaJR3nkeJ356hXrIbMAhx&CITYTOSHIP=dcdcd&COUNTRYTOSHIP=US&AVSDATA=XXN&ACCT=5100&AUTHCODE=333PNI&FIRSTNAME=Oleh&RESULT=0&IAVS=X&POSTFPSMSG=No+Rules+Triggered&BILLTOSTATE=AK&BILLTOCOUNTRY=US&EXPDATE=0725&EMAIL=o.usik%40atwix.com&CARDTYPE=1&PREFPSMSG=No+Rules+Triggered&SHIPTOZIP=12345&PROCAVS=U&COUNTRY=US&AVSZIP=X&ADDRESS=dcdc%40gmail.com&BILLTONAME=Oleh+Usik&ADDRESSTOSHIP=dcdc%40gmail.com&AVSADDR=X&SECURETOKENID=hOidzKMVR3KMFeTS6UpLJVMacQ5GMuhk&SHIPTOSTATE=AK&TRANSTIME=2020-06-21+11%3A36%3A53"
      
    }
  ){
    cart {
      selected_payment_method {
        code
      }
    }
  }
}
</pre>

<img width="1353" alt="Screenshot 2020-06-22 at 13 20 23" src="https://user-images.githubusercontent.com/13517181/85276893-2ad0c400-b48b-11ea-945e-0b7499014ec5.png">

**8.** Place order
<pre> 
mutation {
  placeOrder(
    input: {
      cart_id: "zqYVjMInihkXlKVJAfgbY2yQsVdc6qLR"
    }
  ) {
    order {
      order_number
    }
  }
}
</pre>

<img width="1374" alt="Screenshot 2020-06-22 at 13 21 41" src="https://user-images.githubusercontent.com/13517181/85277051-6e2b3280-b48b-11ea-8392-728dd15a165a.png">

**Result**
<img width="1364" alt="Screenshot 2020-06-22 at 13 23 28" src="https://user-images.githubusercontent.com/13517181/85277211-afbbdd80-b48b-11ea-89b9-36f302d2523c.png">
<img width="780" alt="Screenshot 2020-06-22 at 13 24 02" src="https://user-images.githubusercontent.com/13517181/85277219-b34f6480-b48b-11ea-8f66-f795bf0ce8f3.png">
